conda/models/match_spec.py

```diff
@@ -477,28 +477,42 @@ def _parse_spec_str(spec_str):
 
     # Step 3. strip off brackets portion
     brackets = {}
-    m1 = re.match(r'^(.*)(?:\[(.*)\])$', spec_str)
-    if m1:
-        spec_str, brackets_str = m1.groups()
-        brackets_str = brackets_str.strip("[]\n\r\t ")
-
-        m5 = re.finditer(r'([a-zA-Z0-9_-]+?)=(["\']?)([^\'"]*?)(\2)(?:[, ]|$)', brackets_str)
-        for match in m5:
+    m3 = re.match(r'.*(?:(\[.*\]))', spec_str)
+    if m3:
+        brackets_str = m3.groups()[0]
+        spec_str = spec_str.replace(brackets_str, '')
+        brackets_str = brackets_str[1:-1]
+        m3b = re.finditer(r'([a-zA-Z0-9_-]+?)=(["\']?)([^\'"]*?)(\2)(?:[, ]|$)', brackets_str)
+        for match in m3b:
             key, _, value, _ = match.groups()
             if not key or not value:
                 raise CondaValueError("Invalid MatchSpec: %s" % spec_str)
             brackets[key] = value
 
-    # Step 4. strip off '::' channel and namespace
-    m2 = spec_str.rsplit(':', 2)
-    m2_len = len(m2)
-    if m2_len == 3:
-        channel_str, namespace, spec_str = m2
-    elif m2_len == 2:
-        namespace, spec_str = m2
+    # Step 4. strip off parens portion
+    m4 = re.match(r'.*(?:(\(.*\)))', spec_str)
+    parens = {}
+    if m4:
+        parens_str = m4.groups()[0]
+        spec_str = spec_str.replace(parens_str, '')
+        parens_str = parens_str[1:-1]
+        m4b = re.finditer(r'([a-zA-Z0-9_-]+?)=(["\']?)([^\'"]*?)(\2)(?:[, ]|$)', parens_str)
+        for match in m4b:
+            key, _, value, _ = match.groups()
+            parens[key] = value
+        if 'optional' in parens_str:
+            parens['optional'] = True
+
+    # Step 5. strip off '::' channel and namespace
+    m5 = spec_str.rsplit(':', 2)
+    m5_len = len(m5)
+    if m5_len == 3:
+        channel_str, namespace, spec_str = m5
+    elif m5_len == 2:
+        namespace, spec_str = m5
         channel_str = None
-    elif m2_len:
-        spec_str = m2[0]
+    elif m5_len:
+        spec_str = m5[0]
```

Comment:
This line probably won't be covered yet, and that's fine.  We're not supporting namespaces yet.